index.js

```diff
@@ -1,148 +1,290 @@
       query = new Query()
       this.buildPromise(query, args, key)
-      this.ids.set(key, query)
-    } else if (this.ttl > 0) {
-      onHit()
-      if (currentSecond() - query.cachedOn > this.ttl) {
-        // restart
-        this.buildPromise(query, args, key)
-      }
+      this.dedupes.set(key, query)
     } else {
-      onHit()
+      this.onDedupe(key)
     }
 
     return query.promise
   }
 
-  clear (value) {
+  /**
+   * wrap the original func to sync storage
+   */
+  async wrapFunction (args, key) {
+    const storageKey = this.getStorageKey(key)
+    let data = this.storage.get(storageKey)
+    if (data && typeof data.then === 'function') { data = await data }
+
+    if (data !== undefined) {
+      this.onHit(key)
+      return data
+    }
+
+    this.onMiss(key)
+
+    const result = await this.func(args, key)
+
+    if (this.ttl < 1) {
+      return result
+    }
+
+    if (!this.references) {
+      let p = this.storage.set(storageKey, result, this.ttl)
+      if (p && typeof p.then === 'function') {
+        p = await p
+      }
+      return result
+    }
+
+    let references = this.references(args, key, result)
+    if (references && typeof references.then === 'function') { references = await references }
+    // TODO validate references?
+    await this.storage.set(storageKey, result, this.ttl, references)
+
+    return result
+  }
+
+  buildPromise (query, args, key) {
+    query.promise = this.wrapFunction(args, key)
+
+    // we fork the promise chain on purpose
+    query.promise
+      .then(result => {
+        // clear the dedupe once done
+        this.dedupes.set(key, undefined)
+        return result
+      })
+      // TODO do we want an onError event?
+      .catch(() => {
+        this.dedupes.set(key, undefined)
+        // TODO option to remove key from storage on error?
+        // we may want to relay on cache if the original function got error
+        // then we probably need more option for that
+        const r = this.storage.remove(this.getStorageKey(key))
+        if (r && typeof r.catch === 'function') { r.catch(noop) }
+      })
+  }
+
+  async clear (value) {
+    // TODO validate value?
     if (value) {
       const key = this.getKey(value)
-      this.ids.set(key, undefined)
+      this.dedupes.set(key, undefined)
+      await this.storage.remove(this.getStorageKey(key))
       return
     }
-    this.ids.clear()
+    await this.storage.clear(this.getStorageName())
+    this.dedupes.clear()
+  }
+
+  async get (key) {
+    return this.storage.get(key)
+  }
+
+  async set (key, value, ttl, references) {
+    return this.storage.set(key, value, ttl, references)
+  }
+
+  async invalidate (references) {
+    return this.storage.invalidate(references)
   }
 }
 
 class Query {
   constructor () {
     this.promise = null
-    this.cachedOn = null
   }
 }
 
-function noop () {}
+function noop () { }
 
 module.exports.Cache = Cache
+module.exports.createStorage = createStorage
```

Comment:
I would avoid this kind of factory (if possible), because it makes integration is other modules harder.
Can we combine Cache and createStorage in a single function?

Reply:
how about this?

https://github.com/mercurius-js/cache/pull/48/files#diff-b4fbcddf805309bd1c968fe3284f3f864becc8ba38cfae1826eed9290a583f97R56

we can remove the factory function and pass the params to that function as params